### ISSUE TITLE
"-W error" invalidates interface files

### ISSUE THREAD

**Issue by alexarice** [agda/agda#5151]
This may be intentional, but currently it seems that adding the "-W error" flag to agda causes interface files to need to be remade and in fact the generated interface file differs depending on whether the flag is set. 

This makes creating precompiled agda packages very difficult, as they won't work when this flag is turned off/on. Is there any possible workaround, as I expect rebuilding interface files was intentional.

Edit: This was found because cubical's makefile sets this option, which then means none of the precompiled files can be used anywhere else. This didn't seem to be a thing in agda 2.6.1.2

**Comment by nad** [comment]
Perhaps one could make Agda keep track of whether `-Werror` would have made Agda reject the code. Then code that would not have been rejected could be made compatible with `-Werror`.

**Comment by alexarice** [comment]
> Perhaps one could make Agda keep track of whether `-Werror` would have made Agda reject the code. Then code that would not have been rejected could be made compatible with `-Werror`.

Would this mean that typechecking code with `-Werror` that was built without `-Werror` would cause a rebuild, but not the other way round?

**Comment by alexarice** [comment]
Ah I reread what you said and that makes sense now. This would seem to fix the problem. Another solution could be for `-Werror` to not propagate into libraries, though this is likely harder to implement

**Comment by alexarice** [comment]
>  This didn't seem to be a thing in agda 2.6.1.2

Just checked and this likely is a thing in agda 2.6.1.2 but cubical only added -Werror to their build file since the last version

**Comment by jpoiret** [comment]
I ran into this problem as well when packaging cubical for Guix. IMO the current situation about flag propagation isn't great, whether it be for packagers or even just users of git HEADS: running `make` in cubical would type check the stdlib again, and then next time you load the cubical files in Emacs you'll need to typecheck stdlib yet again.

One solution to this that shouldn't be too hard to implement but a bit wonky maybe would be to only propagate top-level flags for files that don't have a corresponding `.agda-lib` file, and error if you're trying to use flags on a file that has an `.agda-lib`.

An example: suppose `Lib.A` is in some Agda library with some flags set like `--without-K`, and you also have a free-standing `test.agda` file with no `OPTIONS` pragma. Calling `agda --auto-inline test.agda` will not type check `Lib.A` with `-W error`, only `test.agda`. Calling `agda --auto-inline lib/a.agda` would error out, saying that `--auto-inline` needs to be set in the `agda-lib` file.

In the end, that means that feature flags should be the same for developers and users of the library, so `Makefile` and `agda-lib` files need to agree. Regarding `-W error` specifically, I could understand that you don't want to have it on while working on the code, so I don't know whether this solution would be ideal.

WDYT? I can take a stab at implementing this.

**Comment by alexarice** [comment]
`-W error` feels slightly different to something like `--auto-inline` or `--without-K` as my guess is that `-W error` shouldn't fundamentally change the way type checking is done, and only reject more programs at the end. In particular, I would expect if a file had been typechecked with `-W error`, then typechecking it again without `-W error` should be close to a no-op. I'm not sure the same could be said about all flags.

**Comment by jespercockx** [comment]
For `--without-K` the situation is somewhat awkward because it can influence the splitting order of the case trees Agda generates, which can theoretically influence how functions compute (though I haven't seen an example yet where this matters in practice). So a file with `--without-K` should perhaps be rechecked when the flag is removed.

**Comment by alexarice** [comment]
Should probably be noted that a potential half-way solution for this would be to allow the results of different typechecks to be stored, so that if you type check `A.agda` with and without K, you end up with two different `.agdai` files. I believe at the moment the typechecking data is just overwritten when you change flags? This would mean that package managers could just typecheck with both flags (which is not ideal but might be better than nothing)

**Comment by jpoiret** [comment]
> For `--without-K` the situation is somewhat awkward because it can influence the splitting order of the case trees Agda generates, which can theoretically influence how functions compute (though I haven't seen an example yet where this matters in practice). So a file with `--without-K` should perhaps be rechecked when the flag is removed.

But surely if a library is built with `--without-K` and someone uses it in a module with `--with-K`, the old case trees are still valid, right? What I'm proposing is simply a change to how Agda flags are specified + propagated, not to how Agda uses them down the line.

> `-W error` feels slightly different to something like `--auto-inline` or `--without-K` as my guess is that `-W error` shouldn't fundamentally change the way type checking is done, and only reject more programs at the end. In particular, I would expect if a file had been typechecked with `-W error`, then typechecking it again without `-W error` should be close to a no-op. I'm not sure the same could be said about all flags.

Right, maybe I'm hijacking the thread a little bit with this example, but since right now `-W error` does re-type-check everything the workaround I mentioned above would still be applicable.

**Comment by alexarice** [comment]
If I'm understanding your solution correctly, this would allow code which is typechecked with `--without-K` to call code from a library which does not use `--without-K`, and this wouldn't cause an error?

**Comment by jpoiret** [comment]
> If I'm understanding your solution correctly, this would allow code which is typechecked with `--without-K` to call code from a library which does not use `--without-K`, and this wouldn't cause an error?

No, this wouldn't change, what would change is to what files the flags given by a CLI invocation apply. The inter-flag compatibility between files would stay the same.
